inc/zoo/swar/associative_iteration.h

```diff
@@ -503,13 +466,94 @@ constexpr auto halvePrecision(SWAR<NB, T> even, SWAR<NB, T> odd) {
     using S = SWAR<NB, T>;
     static_assert(0 == NB % 2, "Only even lane-bitcounts supported");
     using RV = SWAR<NB/2, T>;
-    constexpr auto HalvingMask = doublingMask<NB/2, T>();
+    constexpr auto HalvingMask = SWAR<NB/2, T>::evenLaneMask();
     auto
         evenHalf = RV{even.value()} & HalvingMask,
         oddHalf = RV{(RV{odd.value()} & HalvingMask).value() << NB/2};
     return evenHalf | oddHalf;
 }
 
+
+template <int NB, typename T> struct MultiplicationResult {
+   SWAR<NB, T> result;
+   SWAR<NB, T> overflow;
+};
+
+template <int NB, typename T>
+constexpr auto
+doublingMultiplication(SWAR<NB, T> multiplicand, SWAR<NB, T> multiplier) {
+   using S = SWAR<NB, T>; using D = SWAR<NB * 2, T>;
+   auto [l_even, l_odd] = doublePrecision(multiplicand);
+   auto [r_even, r_odd] = doublePrecision(multiplier);
+   auto
+       res_even = multiplication_OverflowUnsafe(l_even, r_even),
+       res_odd = multiplication_OverflowUnsafe(l_odd, r_odd);
+   return SWAR_Pair<NB * 2, T>{res_even, res_odd};
+}
+
+template <int NB, typename T>
+constexpr MultiplicationResult<NB, T>
+wideningMultiplication(SWAR<NB, T> multiplicand, SWAR<NB, T> multiplier) {
+   using S = SWAR<NB, T>; using D = SWAR<NB * 2, T>;
+   constexpr auto
+       HalfLane = S::NBits,
+       UpperHalfOfLanes = SWAR<S::NBits, T>::oddLaneMask().value();
+   auto [res_even, res_odd] = doublingMultiplication(multiplicand, multiplier);
+   auto result = halvePrecision(res_even, res_odd);
+   auto
+       over_even = D{(res_even.value() & UpperHalfOfLanes) >> HalfLane},
+       over_odd = D{(res_odd.value() & UpperHalfOfLanes) >> HalfLane};
+   auto upper_lanes_overflow = halvePrecision(over_even, over_odd);
+   return {result, upper_lanes_overflow};
+}
+
+template <int NB, typename T>
+constexpr
+auto saturatedMultiplication(SWAR<NB, T> multiplicand, SWAR<NB, T> multiplier) {
+   using S = SWAR<NB, T>;
+   constexpr auto One = S{S::LeastSignificantBit};
+   auto [result, overflow] = wideningMultiplication(multiplicand, multiplier);
+   auto did_overflow = zoo::swar::greaterEqual(overflow, One);
+   auto lane_mask = did_overflow.MSBtoLaneMask();
+   auto saturated = result | lane_mask;
+   return S{saturated};
+}
+
+
+// TODO(Jamie): Add tests from other PR.
+template<int NB, typename T>
+constexpr auto saturatingExponentiation(
+    SWAR<NB, T> x,
```

Comment:
Absolutely not.
We're not removing the non-saturating exponentiation and provide only the saturating exponentiation.  Don't do that.
**Always** the general operation is pre-requisite for the more specific.

Reply:
good call